### PR TITLE
Add a parallel implementation of to_tf_dataset()

### PR DESCRIPTION
**Pull request by Rocketknight1**
Hey all! Here's a first draft of the PR to add a multiprocessing implementation for `to_tf_dataset()`. It worked in some quick testing for me, but obviously I need to do some much more rigorous testing/benchmarking, and add some proper library tests.

The core idea is that we do everything using `multiprocessing` and `numpy`, and just wrap a `tf.data.Dataset` around the output. We could also rewrite the existing single-threaded implementation based on this code, which might simplify it a bit.

Checklist:
- [X] Add initial draft
- [x] Check that it works regardless of whether the `collate_fn` or dataset returns `tf` or `np` arrays
- [x] Check that it works with `tf.string` return data
- [x] Check indices are correctly reshuffled each epoch
- [x] Make sure workers don't try to initialize a GPU device!!
- [x] Check `fit()` with multiple epochs works fine and that the progress bar is correct
- [x] Check there are no memory leaks or zombie processes
- [x] Benchmark performance
- [x] Tweak params for dataset inference - can we speed things up there a bit?
- [x] Add tests to the library
- [x] Add a PR to `transformers` to expose the `num_workers` argument via `prepare_tf_dataset` (will merge after this one is released)
- [x] Stop TF console spam!! (almost)
- [x] Add a method for creating SHM that doesn't crash if it was left and still linked
- [x] Add a barrier for Py <= 3.7 because it doesn't support SharedMemory
- [x] Support string dtypes by converting them into fixed-width character arrays

### CODE REVIEW

**Comment by Rocketknight1** [comment]
Failing because the test server uses Py3.7 but the `SharedMemory` features require Py3.8! I forgot we still support 3.7 for another couple of months. I'm not sure exactly how to proceed, whether I should leave this PR until then, or just gate the feature behind a version check and skip the tests until the Python version catches up.

File: src/datasets/utils/tf_utils.py
Comment (gante) [inline src/datasets/utils/tf_utils.py:262]: some logic here is shared with `dataset_to_tf`, which could perhaps be shared :p

**Review by gante — COMMENTED** [review]
In general looks good, and I'd be happy to approve the current version. However, I have a design choice question (which may help with the py 3.8 problem).

[disclaimer: superficial knowledge of multiprocessing] `NumpyMultiprocessingGenerator` collects batches in a round robin way, which are sent to the main thread through `SharedMemory` (one per process). This is trickier to implement than a `multiprocessing.pool.Pool`-related strategy, which does not need the shared memory and can optionally ignore batch order to increase throughout (and doesn't need py 3.8). Is there a particular reason behind the choice?

**Review by amyeroberts — COMMENTED** [review]
Thanks for implementing this! Really nice PR - clean and easy to follow ❤️ 

I agree with @gante's comments about abstracting some of the shared logic but otherwise looks great to me 👍  I'll approve after doing a final review once all TODO boxes ticked off

**Comment by lhoestq** [comment]
I haven't played with `NumpyMultiprocessingGenerator` so I can't really help here, but this sounds promising :) Otherwise I think it's also fine to allow `num_workers` only for py>=3.8 for now. You can skip the test on 3.7 and make sure to raise an informative error if someone wants to use `num_workers` with 3.7

**Comment by Rocketknight1** [comment]
Lots of comments here - I'll reply to the specific code comments underneath them, but in response to the general comments:

@gante: I think this approach is much more performant than a `multiprocessing.Pool`. The reason is that when results are returned from a process `Pool`, the returned Python objects are pickled by the child processes, sent down a pipe and unpickled by the parent process. This creates a huge single-process bottleneck as the parent has to unpickle lots of large NumPy arrays, which is quite slow.

When you use a `SharedMemory` approach, the data is just **there** for the parent process - the child and the parent are writing to exactly the same array in memory, and no pickling or unpickling occurs. This means the parent can just immediately copy the array (which is much faster than unpickling) and yield it to `tf.data`. We're taking advantage of the fact that we know the data is just big NumPy arrays and we don't need the full generality of `pickle`.

@lhoestq: Sounds good! I'll add a clear error and skip the tests on Py<=3.7.

File: src/datasets/utils/tf_utils.py
Comment (Rocketknight1) [reply src/datasets/utils/tf_utils.py:235]: Yes - the child processes use negative shape values to indicate that they're finished and no more data is coming! I should definitely comment this more clearly.

File: src/datasets/utils/tf_utils.py
Comment (Rocketknight1) [reply src/datasets/utils/tf_utils.py]: Fixed!

File: src/datasets/utils/tf_utils.py
Comment (Rocketknight1) [reply src/datasets/utils/tf_utils.py:107]: I think so! You can't index into the output of `dict.values()`, and I don't know of a better way to get a single arbitrary element from a dict (you can use `popitem()` but that deletes the value from the dict)

File: src/datasets/utils/tf_utils.py
Comment (Rocketknight1) [reply src/datasets/utils/tf_utils.py]: Done!

File: tests/test_arrow_dataset.py
Comment (Rocketknight1) [reply tests/test_arrow_dataset.py:2637]: You're right, that line is never triggered. My bad!

**Review by gante — APPROVED** [review]
LGTM 👍 (and thanks for elaborating on the `Pool` vs `SharedMemory`!)

**Comment by Rocketknight1** [comment]
Also, an extra technicality, just for information in case anyone looks at this PR later: Recent versions of Python allow [pickled objects to store out-of-band data](https://peps.python.org/pep-0574/). This allows for very efficient zero-copy unpickling of objects like NumPy arrays, with the unpickled object having a view on the same memory as the original. 

However, this explicitly does **not** work when the object is unpickled by a different process than the one that created it. For this to work you must explicitly allocate shared memory and create the array there, which pickle cannot handle for you. As a result, if you just benchmark unpickling vs copying of NumPy arrays it can seem like unpickling is very fast - but this is only true when the pickle was created in the unpickling process!

File: src/datasets/arrow_dataset.py
Comment (amyeroberts) [inline src/datasets/arrow_dataset.py:433]: Bit picky, but I would do something like (unless a negative number indicates something special?)

```
if num_workers == 0:
    ...
elif num_workers >= 1:
    ...
else:
    raise ValueError(f"Invalid number of workers {num_workers}. Must be >= 0")
```

**Review by amyeroberts — APPROVED** [review]
LGTM! Thanks for working on this. Looking forward to seeing it in action 🏎️ 💨

File: src/datasets/arrow_dataset.py
Comment (Rocketknight1) [reply src/datasets/arrow_dataset.py:433]: Haha, no, this time negatives do not have special meaning! Your approach is better, I'll make the change

**Comment by Rocketknight1** [comment]
@lhoestq @amyeroberts @gante I did a substantial rewrite and all tests are passing now (Windows seems to time out or something and I can't figure out why - not sure if that's related to this PR!). I also confirmed tests are passing locally with Py==3.10. 

Aside from incorporating everyone's comments, I also made a context manager to create and handle shared memory - this ensures that shared memory is cleaned up even if execution is interrupted. Also, shared memory names include a UUID string now to avoid collisions. Finally, string arrays are now split up into fixed-width character arrays in the workers so that they can be passed through a shared array, and the parent process reconstructs them into string arrays.

**Comment by Rocketknight1** [comment]
Update: `test_arrow_dataset.py` ran fine in this branch on my Windows machine (Py 3.10), so I have no idea what's up with this

**Review by lhoestq — COMMENTED** [review]
Nice ! Thanks for adding comments to explain the most tricky parts :)

It looks good to me, feel free to just add more type hints and address my minor comments

File: src/datasets/utils/tf_utils.py
Comment (Rocketknight1) [reply src/datasets/utils/tf_utils.py:194]: Whoops, sorry - those are definitely not supposed to be in the final version!

File: src/datasets/utils/tf_utils.py
Comment (Rocketknight1) [reply src/datasets/utils/tf_utils.py:132]: Done!

File: src/datasets/utils/tf_utils.py
Comment (Rocketknight1) [reply src/datasets/utils/tf_utils.py]: Done!

File: src/datasets/utils/tf_utils.py
Comment (lhoestq) [inline src/datasets/utils/tf_utils.py:478]: can you also put the docstring here ?

File: src/datasets/utils/tf_utils.py
Comment (lhoestq) [inline src/datasets/utils/tf_utils.py:497]: Does using from_generator have consequences on how the dataset can be used compared to from_tensor_slices ?
If so, this must be documented

**Review by lhoestq — COMMENTED** [review]
Thanks ! Before merging, can you update the use_with_tensorflow.mdx documentation page ?

File: src/datasets/utils/tf_utils.py
Comment (Rocketknight1) [reply src/datasets/utils/tf_utils.py:497]: No, from a UX perspective the output dataset should be identical!

File: src/datasets/utils/tf_utils.py
Comment (Rocketknight1) [reply src/datasets/utils/tf_utils.py:478]: Done!

**Comment by Rocketknight1** [comment]
Tests seem to be failing for unrelated reasons.

**Comment by lhoestq** [comment]
Tests are failing because of a bug on the Hub side - this is being fixed :)

lmk once the TF documentation page is updated and we can merge !

**Comment by Rocketknight1** [comment]
@lhoestq Docs updated!

**Review by lhoestq — APPROVED** [review]
Great ! LGTM :)

**Pull request merged**
